### DISCUSSION TITLE
Access to the current database

### DISCUSSION THREAD

**Discussion by piotrklibert** [Deleplace/programming-idioms#207] [General]
Hi,

as a side project, I'm developing a desktop app that uses "spaced repetition" to help with learning programming languages. To use that, you would need a collection of "cards" - each with a question on one side and an answer on the other. While everybody is bound to need different stacks of cards, I want to have some as an example.

I thought about Rosetta Code, but there the examples are much longer, and I don't want to scrape the Wiki (it's painful). Programming Idioms, on the other hand, are short enough to fit on a single card.

My question is: can I get direct access to the current database of code snippets on PI? It could be exported to just about any format, I'll manage. Or if there's an API that I could call, that would also be ok. Is any of these possible?

If it matters, I plan to release the app as GPL.

**Comment by Deleplace** [comment]
Hi Piotr,
here is a JSON dump. It is an array of "idioms", each containing an array of "implementations".

The contents are [CC-BY-SA](https://en.wikipedia.org/wiki/Wikipedia:Text_of_Creative_Commons_Attribution-ShareAlike_3.0_Unported_License) so yes how you would use it does matter. GPL should be fine, please make sure to keep the attribution to the author of the idiom and the author of the implementation.

Note that:
- each snippet is just "one way of doing the job", it should not be regarded as the only source of truth,
- an idiom may have several implementations for the same language,
- the implementations are contributed by the community, so the quality is not guaranteed, some of them are imperfect in many ways.

[programming-idioms.org.2022-07-15_11-59.json.zip](https://github.com/Deleplace/programming-idioms/files/9120573/programming-idioms.org.2022-07-15_11-59.json.zip)

**Reply by piotrklibert** [reply]
Thanks a lot! Once I have enough to make a screenshot, I'll come back to you to make sure it includes all the licensing info needed.

As for the dataset: I used it earlier this year when I was learning Go and Rust, and also contributed a bunch of solutions for Smalltalk recently. I'm actually a polyglot programmer - out of languages included on PI there are only a few that I don't know: VB, Cobol, Fortran, and Perl5 (but I know Raku/Perl6). I think I will be able to work with the dataset and select examples that would be the most helpful for learning a language. I'm currently in the process of learning Kotlin, so I will start with it.

Thanks again, I'll let you know once I have a proof-of-concept/MVP.

**Reply by Deleplace** [reply]
Awesome. Have fun building the cards

**Reply by piotrklibert** [reply]
Well, it took some time due to reasons, but I finally have something screenshot-worthy:
![screenshot_2023-01-09_1244](https://user-images.githubusercontent.com/1244294/211300907-f999c852-35eb-431f-a982-65ba853e2ed2.png)
I'm not done yet, but I wanted to share a bit of the progress that I made :)

**Reply by Deleplace** [reply]
Cool!
The underscores are lightweight formatting, and the token following "\_" is displayed as emphasized in the website. You may want to display plain text ([stripping](https://programming-idioms.org/idiom/167/trim-prefix) the "\_" character), or to apply an emphasis style.

The code snippets don't contain underscored styling. Only the idiom title, idiom statement, and implementation comments have them.

**Reply by piotrklibert** [reply]
I'm probably going to replace `_` by surrounding the identifier in `*` for now. Currently, I split the JSON into a bunch of markdown files, then used `pygments` to render those files to the png files I'm displaying. Fortunately, `pygments` is capable of syntax-highlighting code blocks inside markdown. But this is just a temporary measure to get something to display with the least amount of effort: I'm going to render the description and the code separately, and then I'll render the underscores in descriptions properly as bold. I'm not sure right now how to get both syntax highlighting and make selected identifiers bold at the same time, will need to figure it out.

I forgot about the impl comments! I need to render those somehow, too. I also still need to write the logic for choosing the right snippet to display (currently, it's just random) and add a few buttons to let the user say how well they think they remember a given snippet. I'll get back here with an update once I have that in place and pushed to GH :)

**Reply by Deleplace** [reply]
You don't need to mix syntax highlighting with bold identifiers:
- imports and code need syntax highlighting only,
- title, lead, paragraph, comments need bold identifiers only.

**Reply by piotrklibert** [reply]
> imports and code need syntax highlighting only,

That's the point - I want the symbols named in a task description to be distinct from other symbols in the highlighted code. Why not? They could be underlined, for example. However, doing that is problematic with my current approach to generating cards, so I decided to leave this for later.

I put the rendered cards here: https://klibert.pl/statics/cards/ - I get to see one of them every 7 minutes for 20 seconds, avoiding repetitions. It's much less distracting than I thought, and I have learned quite a lot about Dart in the last few days. 

I also added links to PI edit pages at the top:
![screenshot_2023-01-16_2305](https://user-images.githubusercontent.com/1244294/212772841-0b1875e3-60c7-4200-a2f5-c482776a54d9.png)

which made me fix some Python and add one Lua implementation. It's fun! :)

**Comment by piotrklibert** [comment]
@Deleplace I'd like to add Raku (formerly known as Perl6, up to 2019) implementations, but when I tried I got a message saying Raku is not currently supported. What needs to happen for Raku to be available?

**Comment by Deleplace** [comment]
If #90 gets a lot of traction, Raku will be added. Note that it is not a goal to support all languages.

**Reply by piotrklibert** [reply]
Yes, supporting **all** languages would be utterly impossible; there are easily hundreds of langs that are actually used by someone, and thousands of toy or experimental ones.

Gaining traction in the Raku case is going to be hard, unfortunately, as the community is very small... It's a shame because Raku is one of the most interesting languages I have encountered. Still, Cobol and Fortran are supported, so maybe there's hope. I could prepare around 100 idiom implementations to start with - would that help?
